File: src/decoder/ac3-decoder.ts
/*
 * AC-3 (Dolby Digital) Audio Decoder
 *
 * TypeScript wrapper for liba52 WASM decoder
 */

import {
  AudioDecoder,
  AudioDecoderInfo,
  WasmModule,
} from "./wasm-audio-decoder";

// AC-3 constants
const AC3_SAMPLES_PER_FRAME = 1536; // 6 blocks * 256 samples
const AC3_MAX_CHANNELS = 6;
const AC3_HEADER_SIZE = 7; // Minimum bytes needed for syncinfo

// Sync info array indices
const SYNC_FLAGS = 0;
const SYNC_SAMPLE_RATE = 1;
const SYNC_BIT_RATE = 2;
const SYNC_FRAME_SIZE = 3;
const SYNC_INFO_SIZE = 4;

// Decode info array indices
const DECODE_SAMPLES = 0;
const DECODE_SAMPLE_RATE = 1;
const DECODE_CHANNELS = 2;
const DECODE_INFO_SIZE = 3;

export class AC3Decoder implements AudioDecoder {
  readonly name = "AC3Decoder";

  private module: WasmModule | null = null;
  private decoderPtr: number = 0;
  private inputPtr: number = 0;
  private outputPtr: number = 0;
  private syncInfoPtr: number = 0;
  private decodeInfoPtr: number = 0;
  private inputBufferSize: number = 0;
  private _ready: Promise<void>;
  private _isReady: boolean = false;

  constructor(wasmModuleFactory: () => Promise<WasmModule>) {
    this._ready = this.init(wasmModuleFactory);
  }

  get ready(): Promise<void> {
    return this._ready;
  }

  get isReady(): boolean {
    return this._isReady;
  }

  private async init(wasmModuleFactory: () => Promise<WasmModule>): Promise<void> {
    this.module = await wasmModuleFactory();

    // Create decoder instance
    const create = this.module.cwrap("ac3_decoder_create", "number", []);
    this.decoderPtr = create();

    if (!this.decoderPtr) {
      throw new Error("Failed to create AC-3 decoder");
    }

    // Allocate output buffer (max samples * max channels * 4 bytes per float)
    const outputSize = AC3_SAMPLES_PER_FRAME * AC3_MAX_CHANNELS * 4;
    this.outputPtr = this.module._malloc(outputSize);

    // Allocate info buffers
    this.syncInfoPtr = this.module._malloc(SYNC_INFO_SIZE * 4);
    this.decodeInfoPtr = this.module._malloc(DECODE_INFO_SIZE * 4);

    // Input buffer allocated on demand
    this.inputBufferSize = 0;
    this.inputPtr = 0;

    this._isReady = true;
  }

  /**
   * Get sync info from AC-3 frame header
   *
   * @param input - Input buffer (needs at least 7 bytes)
   * @returns Frame info or null if not a valid AC-3 sync
   */
  getSyncInfo(input: Uint8Array): {
    flags: number;
    sampleRate: number;
    bitRate: number;
    frameSize: number;
  } | null {
    if (!this._isReady || !this.module || input.length < AC3_HEADER_SIZE) {
      return null;
    }

    // Ensure input buffer is large enough
    if (input.length > this.inputBufferSize) {
      if (this.inputPtr) {
        this.module._free(this.inputPtr);
      }
      this.inputBufferSize = Math.max(input.length, 4096);
      this.inputPtr = this.module._malloc(this.inputBufferSize);
    }

    // Copy input data to WASM memory
    this.module.HEAPU8.set(input, this.inputPtr);

    // Get sync info
    const syncinfo = this.module.cwrap(
      "ac3_syncinfo",
      "number",
      ["number", "number"],
    );

    const frameSize = syncinfo(this.inputPtr, this.syncInfoPtr);

    if (frameSize <= 0) {
      return null;
    }

    return {
      flags: this.module.getValue(this.syncInfoPtr + SYNC_FLAGS * 4, "i32"),
      sampleRate: this.module.getValue(this.syncInfoPtr + SYNC_SAMPLE_RATE * 4, "i32"),
      bitRate: this.module.getValue(this.syncInfoPtr + SYNC_BIT_RATE * 4, "i32"),
      frameSize: this.module.getValue(this.syncInfoPtr + SYNC_FRAME_SIZE * 4, "i32"),
    };
  }

  /**
   * Decode AC-3 frame
   *
   * @param input - Input buffer containing complete AC-3 frame
   * @returns Decoded PCM data and frame info, or null on error
   */
  decode(input: Uint8Array): { pcm: Float32Array; info: AudioDecoderInfo } | null {
    if (!this._isReady || !this.module) {
      return null;
    }

    // Check for valid sync
    const syncInfo = this.getSyncInfo(input);
    if (!syncInfo || input.length < syncInfo.frameSize) {
      return null;
    }

    // Copy input data to WASM memory (may have been resized in getSyncInfo)
    this.module.HEAPU8.set(input, this.inputPtr);

    // Decode frame
    const decode = this.module.cwrap(
      "ac3_decode_frame",
      "number",
      ["number", "number", "number", "number", "number"],
    );

    const samples = decode(
      this.decoderPtr,
      this.inputPtr,
      input.length,
      this.outputPtr,
      this.decodeInfoPtr,
    );

    if (samples <= 0) {
      return null;
    }

    // Read decode info
    const decodedSamples = this.module.getValue(
      this.decodeInfoPtr + DECODE_SAMPLES * 4,
      "i32",
    );
    const sampleRate = this.module.getValue(
      this.decodeInfoPtr + DECODE_SAMPLE_RATE * 4,
      "i32",
    );
    const channels = this.module.getValue(
      this.decodeInfoPtr + DECODE_CHANNELS * 4,
      "i32",
    );

    // Read output samples (float32, already interleaved)
    const outputSamples = decodedSamples * channels;
    const pcm = new Float32Array(
      this.module.HEAPF32.buffer,
      this.outputPtr,
      outputSamples,
    ).slice(); // Copy to avoid referencing WASM memory

    return {
      pcm,
      info: {
        samples: decodedSamples,
        sampleRate,
        channels,
      },
    };
  }

  /**
   * Decode multiple frames from a buffer
   */
  decodeAll(input: Uint8Array): { pcm: Float32Array; info: AudioDecoderInfo } | null {
    if (!this._isReady || !this.module) {
      return null;
    }

    const pcmChunks: Float32Array[] = [];
    let totalSamples = 0;
    let lastInfo: AudioDecoderInfo | null = null;
    let offset = 0;

    while (offset < input.length) {
      const remaining = input.subarray(offset);

      // Get frame size
      const syncInfo = this.getSyncInfo(remaining);
      if (!syncInfo) {
        // Try to find next sync word
        offset++;
        continue;
      }

      if (remaining.length < syncInfo.frameSize) {
        break;
      }

      const frameData = remaining.subarray(0, syncInfo.frameSize);
      const result = this.decode(frameData);

      if (!result) {
        offset += syncInfo.frameSize;
        continue;
      }

      pcmChunks.push(result.pcm);
      totalSamples += result.info.samples;
      lastInfo = result.info;
      offset += syncInfo.frameSize;
    }

    if (pcmChunks.length === 0 || !lastInfo) {
      return null;
    }

    // Concatenate all PCM chunks
    const pcm = new Float32Array(totalSamples * lastInfo.channels);
    let writeOffset = 0;
    for (const chunk of pcmChunks) {
      pcm.set(chunk, writeOffset);
      writeOffset += chunk.length;
    }

    return {
      pcm,
      info: {
        samples: totalSamples,
        sampleRate: lastInfo.sampleRate,
        channels: lastInfo.channels,
      },
    };
  }

  /**
   * Reset decoder state
   */
  reset(): void {
    // AC-3 decoder doesn't have persistent state that needs reset
    // Each frame is independently decodable
  }

  /**
   * Destroy decoder and free all resources
   */
  destroy(): void {
    if (!this.module) {
      return;
    }

    if (this.decoderPtr) {
      const destroy = this.module.cwrap(
        "ac3_decoder_destroy",
        null,
        ["number"],
      );
      destroy(this.decoderPtr);
      this.decoderPtr = 0;
    }

    if (this.inputPtr) {
      this.module._free(this.inputPtr);
      this.inputPtr = 0;
    }

    if (this.outputPtr) {
      this.module._free(this.outputPtr);
      this.outputPtr = 0;
    }

    if (this.syncInfoPtr) {
      this.module._free(this.syncInfoPtr);
      this.syncInfoPtr = 0;
    }

    if (this.decodeInfoPtr) {
      this.module._free(this.decodeInfoPtr);
      this.decodeInfoPtr = 0;
    }

    this.module = null;
    this._isReady = false;
  }
}

File: wasm/liba52/ac3_decoder.c
/*
 * AC-3 Audio Decoder - WebAssembly Wrapper
 * 
 * This file provides a simple C interface for decoding AC-3 (Dolby Digital)
 * audio using liba52, exported as WebAssembly functions.
 * 
 * Copyright (C) 2026 mpegts.js contributors
 * Licensed under GPL-2.0 (same as liba52)
 */

#include "src/config.h"
#include "include/a52.h"
#include "src/a52_internal.h"

#ifdef __EMSCRIPTEN__
#include <emscripten.h>
#define EXPORT EMSCRIPTEN_KEEPALIVE
#else
#define EXPORT
#endif

#include <stdlib.h>
#include <string.h>

/* AC-3 frame always has 6 blocks of 256 samples = 1536 samples per channel */
#define AC3_SAMPLES_PER_FRAME 1536
#define AC3_MAX_CHANNELS 6

/* Decoder wrapper structure */
typedef struct {
    a52_state_t *state;
    sample_t *samples;
    int initialized;
} AC3Decoder;

/*
 * Create a new AC-3 decoder instance
 * Returns: pointer to decoder state, or NULL on failure
 */
EXPORT
AC3Decoder* ac3_decoder_create(void) {
    AC3Decoder* decoder = (AC3Decoder*)malloc(sizeof(AC3Decoder));
    if (!decoder) {
        return NULL;
    }
    
    /* Initialize liba52 with no SIMD acceleration */
    decoder->state = a52_init(0);
    if (!decoder->state) {
        free(decoder);
        return NULL;
    }
    
    decoder->samples = a52_samples(decoder->state);
    decoder->initialized = 1;
    
    return decoder;
}

/*
 * Destroy decoder instance and free memory
 */
EXPORT
void ac3_decoder_destroy(AC3Decoder* decoder) {
    if (decoder) {
        if (decoder->state) {
            a52_free(decoder->state);
        }
        free(decoder);
    }
}

/*
 * Get sync info from an AC-3 frame header
 * 
 * Parameters:
 *   buf         - input buffer (needs at least 7 bytes for sync info)
 *   out_info    - output array: [flags, sample_rate, bit_rate, frame_size]
 * 
 * Returns: frame size in bytes, or 0 if not a valid sync
 */
EXPORT
int ac3_syncinfo(const unsigned char* buf, int* out_info) {
    int flags, sample_rate, bit_rate;
    int frame_size = a52_syncinfo((uint8_t*)buf, &flags, &sample_rate, &bit_rate);
    
    if (frame_size > 0 && out_info) {
        out_info[0] = flags;
        out_info[1] = sample_rate;
        out_info[2] = bit_rate;
        out_info[3] = frame_size;
    }
    
    return frame_size;
}

/*
 * Decode a complete AC-3 frame
 * 
 * Parameters:
 *   decoder     - decoder instance
 *   input       - input buffer containing complete AC-3 frame
 *   input_size  - size of input buffer (should match frame_size from syncinfo)
 *   output      - output buffer for interleaved float PCM samples
 *                 (must be at least AC3_SAMPLES_PER_FRAME * channels * sizeof(float))
 *   out_info    - output array: [samples_decoded, sample_rate, channels]
 * 
 * Returns: number of samples decoded per channel, or negative on error
 */
EXPORT
int ac3_decode_frame(
    AC3Decoder* decoder,
    const unsigned char* input,
    int input_size,
    float* output,
    int* out_info
) {
    if (!decoder || !decoder->initialized || !input || !output || !out_info) {
        return -1;
    }
    
    int flags, sample_rate, bit_rate;
    int frame_size = a52_syncinfo((uint8_t*)input, &flags, &sample_rate, &bit_rate);
    
    if (frame_size == 0 || frame_size > input_size) {
        return -2; /* Invalid or incomplete frame */
    }
    
    /* Determine output configuration - downmix to stereo for simplicity */
    int output_flags = A52_STEREO;
    sample_t level = 1.0f;
    sample_t bias = 0.0f;
    
    /* Parse the frame */
    if (a52_frame(decoder->state, (uint8_t*)input, &output_flags, &level, bias)) {
        return -3; /* Frame parse error */
    }
    
    /* Get actual channel count from output flags */
    int channels = 2; /* We're downmixing to stereo */
    
    /* Decode all 6 blocks */
    int total_samples = 0;
    for (int i = 0; i < 6; i++) {
        if (a52_block(decoder->state)) {
            return -4; /* Block decode error */
        }
        
        /* Copy interleaved samples to output */
        sample_t* samples = decoder->samples;
        for (int s = 0; s < 256; s++) {
            for (int c = 0; c < channels; c++) {
                output[(total_samples + s) * channels + c] = samples[c * 256 + s];
            }
        }
        total_samples += 256;
    }
    
    /* Fill output info */
    out_info[0] = total_samples;  /* samples per channel */
    out_info[1] = sample_rate;
    out_info[2] = channels;
    
    return total_samples;
}

/*
 * Get the number of samples per AC-3 frame
 */
EXPORT
int ac3_samples_per_frame(void) {
    return AC3_SAMPLES_PER_FRAME;
}

/*
 * Get the maximum number of channels
 */
EXPORT
int ac3_max_channels(void) {
    return AC3_MAX_CHANNELS;
}
